pkg/operation/botanist/etcdencryption.go

```diff
@@ -0,0 +1,321 @@
+		b.Logger.Info("no EncryptionConfiguration found in seed. Creating new passive configuration.")
+		ecSeed, err = encryptionconfiguration.CreateNewPassiveConfiguration()
+		if err != nil {
+			return err
+		}
+		needToWriteConfig = true
+	} else {
+		exists, ecGarden, err := b.readEncryptionConfigurationFromGarden()
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return fmt.Errorf("EncryptionConfiguration exists in seed but not in garden")
+		}
+		// if it exists already, it needs to be consistent
+		consistent, err := b.isEncryptionConfigurationConsistent(ecSeed, ecGarden)
+		if err != nil {
+			return err
+		}
+		if !consistent {
+			return fmt.Errorf("EncryptionConfiguration in seed is not consistent")
+		}
+		if forcePlaintextSecrets {
+			if encryptionconfiguration.IsActive(ecSeed) {
+				// if annotation requires plaintext secrets, deactivate configuration if it is still active
+				b.Logger.Infof("EncryptionConfiguration is annotated with %v. De-activating existing active EncryptionConfiguration. Note: shoot secrets will be decrypted in etcd subsequently", common.EtcdEncryptionForcePlaintextAnnotationName)
+				encryptionconfiguration.SetActive(ecSeed, false)
+				needToWriteConfig = true
+			}
+		} else if !encryptionconfiguration.IsActive(ecSeed) {
+			// if it is not active already (meaning that identity and not aescbc
+			// is still the first encryption provider) then set it to active
+			// and remember to write this created configuration
+			b.Logger.Info("activating existing passive EncryptionConfiguration")
+			encryptionconfiguration.SetActive(ecSeed, true)
+			needToWriteConfig = true
+		}
+	}
+	ecSeedYamlBytes, err := encryptionconfiguration.ToYAML(ecSeed)
+	if err != nil {
+		return err
+	}
+	if needToWriteConfig {
+		b.Logger.Info("writing new/updated EncryptionConfiguration")
+		if err := b.writeEncryptionConfiguration(ecSeedYamlBytes, forcePlaintextSecrets); err != nil {
+			return err
+		}
+	}
+
+	checksum := utils.ComputeSHA256Hex(ecSeedYamlBytes)
+	b.mutex.Lock()
+	b.CheckSums[common.EtcdEncryptionSecretName] = checksum
+	b.mutex.Unlock()
+	// enablement of etcd encryption feature done in helm chart of apiserver deployment
+	return nil
+}
+
+// RewriteShootSecretsIfEncryptionConfigurationChanged rewrites a shoot's secrets if the EncryptionConfiguration has changed
+func (b *Botanist) RewriteShootSecretsIfEncryptionConfigurationChanged(ctx context.Context) error {
+	// WARNING:
+	// No explicit checking of whether EncryptionConfiguration is contained in a backup.
+	// Be aware of the risk!
+	//
+	// TODO: Ensure this is also agreed upon by Gardener team
+
+	// TODO: contact Amshuman Rao Karaya
+
+	enableEtcdEncryption, err := utils.CheckVersionMeetsConstraint(b.Shoot.Info.Spec.Kubernetes.Version, ">= 1.13")
+	if err != nil {
+		return err
+	}
+	if !enableEtcdEncryption {
+		return nil
+	}
+	needToRewrite, err := b.needToRewriteShootSecrets()
+	if err != nil {
+		return err
+	}
+	if needToRewrite {
```

Comment:
Classically, we try to exit early. That means that the "business logic" of the function is at the very end and only executed after all checks passed.
So, I suggest to reverse the flow:

```suggestion
	if !needToRewrite {
		return nil
	}
	
	b.Logger.Info("rewriting shoot secrets")
	return b.rewriteShootSecrets()
```